d2/src/test/java/com/linkedin/d2/balancer/simple/SimpleLoadBalancerTest.java

```diff
@@ -439,117 +439,81 @@ public void testClusterInfoProviderUnregisterClusterListener()
   }
 
   @Test
-  public void testListenToServiceAndClusterTimeout() throws Exception
+  public void testListenToServiceAndClusterTimeout()
   {
     MockStore<ServiceProperties> serviceRegistry = new MockStore<>();
     MockStore<ClusterProperties> clusterRegistry = new MockStore<>();
     MockStore<UriProperties> uriRegistry = new MockStore<>();
-    LoadBalancerTestState state = spy(new LoadBalancerTestState());
-    state.listenToService = false;
-    SimpleLoadBalancer loadBalancer = spy(setupLoadBalancer(state, serviceRegistry, clusterRegistry, uriRegistry));
+    SimpleLoadBalancerState state =
+        spy(new SimpleLoadBalancerState(new SynchronousExecutorService(), uriRegistry, clusterRegistry, serviceRegistry,
+                                        new HashMap<>(), new HashMap<>()));
+    doAnswer(invocation ->
+             {
+               Thread.sleep(100);
+               return null;
+             }).when(state).listenToService(any(), any());
+    SimpleLoadBalancer loadBalancer = spy(new SimpleLoadBalancer(state, 1, TimeUnit.MILLISECONDS, _d2Executor));
     // case1: listenToService timeout, and simpleLoadBalancer not hit the cache value
-    Callback<ServiceProperties> callback = new Callback<ServiceProperties>()
-    {
-      @Override
-      public void onError(Throwable e)
-      {
-        assertTrue(e instanceof ServiceUnavailableException);
-      }
-
-      @Override
-      public void onSuccess(ServiceProperties result)
-      {
-        throw new RuntimeException("onSuccess should not be called");
-      }
-    };
+    FutureCallback<ServiceProperties> callback = spy(new FutureCallback<>());
```

Comment:
Why not use a mock directly rather than building a spy?

Reply:
Please see the latest code. It needs the `spy` for using `callback.get()` method. Thanks for ur suggestion.

Reply:
Spy an object is needed when mocking some method's result (while other methods do the real behavior). Simply calling .get method on the callback shouldn't need it.